### PR TITLE
Added real-time graphing visualizer for samplers

### PR DESCRIPTION
**Pull request by S-Ptr**

### CODE REVIEW

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.jmeter.visualizers.gui.AbstractVisualizer;
+import org.knowm.xchart.QuickChart;
+import org.knowm.xchart.XChartPanel;
+import org.knowm.xchart.XYChart;
+
+/**
+ *
+ * @author User
+ */
+public class FrejaGraphVisualizerGui extends AbstractVisualizer {
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: Rename to FrejaEidResultsGraphVisualizerGui

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
+/*
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: Remove this commented part

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    HashMap<String, GraphPanel> panelList;
+    private static final String RESPONSE_CODE = "FAILED";
+
+    public FrejaGraphVisualizerGui() {
+        super();
+        panelList = new HashMap<>();
+        setLayout(new BorderLayout(0, 5));
+        setBorder(makeBorder());
+        add(makeTitlePanel(), BorderLayout.NORTH);
+        authResults = new GraphPanel("Authentication Sampler Results", "Number" ,"Count");
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: move this to separate private method called createFrejaEidResultsPanels() or something like this.

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        signResults = new GraphPanel("Sign Sampler Results", "Number" ,"Count");
+        openSecureResults = new GraphPanel("Open Secure Connection Results", "Number" ,"Count");
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: Rename this to openSecureConnectionResults

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        jTP.add("Auth", authResults.getPanel());
+        jTP.add("Sign", signResults.getPanel());
+        jTP.add("Open secure connection", openSecureResults.getPanel());
+        add(jTP, BorderLayout.CENTER);
+
+    }
+
+    @Override
+    public String getStaticLabel() {
+        return "Freja eID Graph";
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: Rename to Freja eID Results Graph

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            try {
+                bais.close();
+            } catch (IOException ex) {
+                Logger.getLogger(FrejaEidPluginVisualizerGui.class.getName()).log(Level.SEVERE, null, ex);
+            }
+        }
+        return responseData;
+    }
+    
+    private void statisticsOneRequest(String request, String responseCode) {
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: rename to statisticsForOneRequest

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        GraphPanel result = panelList.get(request);
+        if (responseCode.equals(RESPONSE_CODE)) {
+            result.increaseFailed();
+        } else {
+            result.increaseDelivered();
+        }
+    }
+
+    private void statisticsMoreRequests(SampleResult sampleResult) {
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: rename statisticsForRequests

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+import org.knowm.xchart.QuickChart;
+import org.knowm.xchart.XChartPanel;
+import org.knowm.xchart.XYChart;
+
+/**
+ *
+ * @author User
+ */
+public class FrejaGraphVisualizerGui extends AbstractVisualizer {
+    List<Double> latency;
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaGraphVisualizerGui.java]: Do we need all this variables here? They can be declared as local method variables, or at least some of them.

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        count = new CopyOnWriteArrayList<>();
+        count.add((double)0);
+    }
+
+    public XChartPanel getPanel() {
+        return chartPanel;
+    }
+
+    /*public void updateSeries(String seriesName, List<? extends Number> xData, List<? extends Number> yData) {
+        chart.updateXYSeries(seriesName, xData, yData, null);
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: Remove this

**Review by risticivan12 — APPROVED** [review]

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.concurrent.CopyOnWriteArrayList;
+import org.knowm.xchart.QuickChart;
+import org.knowm.xchart.XChartPanel;
+import org.knowm.xchart.XYChart;
+
+/**
+ *
+ * @author User
+ */
+public class GraphPanel {
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: Rename FrejaEidResultsGraphPanel

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: Remove this

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    private XChartPanel chartPanel;
+    List<Double> successCount;
+    List<Double> failCount;
+    List<Double> count;
+
+    public GraphPanel(String title, String xAxisName, String yAxisName) {
+        double[] initData = new double[1];
+        initData[0] = 0;
+        chart = QuickChart.getChart(title, xAxisName, yAxisName,"Failed", initData, initData);
+        chart.addSeries("Delivered", initData, initData);
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: Move this chart operations in separate private method describing what is done here.

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * @author User
+ */
+public class GraphPanel {
+
+    private XYChart chart;
+    private XChartPanel chartPanel;
+    List<Double> successCount;
+    List<Double> failCount;
+    List<Double> count;
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: Describe this count to be more descriptive. If it is summary of successCount and failCount, describe it like that.

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        chart.updateXYSeries(seriesName, xData, yData, null);
+    }
+
+    public void updateSeries(String seriesName, double[] xData, double[] yData) {
+        chart.updateXYSeries(seriesName, xData, yData, null);
+    }*/
+
+    public void increaseFailed() {
+        count.add(count.get(count.size() - 1) + 1);
+        //get the last element in the list
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: remov comments and describe variables in that way that you don't need comments in code

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        //update the other graph as well, for consistency. No increments here.
+        successCount.add(successCount.get(successCount.size() - 1));
+        chart.updateXYSeries("Failed", count, failCount, null);
+        chart.updateXYSeries("Delivered", count, successCount, null);
+        chartPanel.repaint();
+    }
+
+    public void increaseDelivered() {
+        count.add(count.get(count.size() - 1) + 1);
+        //same deal with successCount as with failCount
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: same as comment above

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+    public void increaseFailed() {
+        count.add(count.get(count.size() - 1) + 1);
+        //get the last element in the list
+        double last = failCount.get(failCount.size() - 1);
+        //append the incremented value of the last element to the list.
+        failCount.add(last + 1);
+        //update the other graph as well, for consistency. No increments here.
+        successCount.add(successCount.get(successCount.size() - 1));
+        chart.updateXYSeries("Failed", count, failCount, null);
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: move this updating of series in private method and describe it, reuse it here

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+        chartPanel.repaint();
+    }
+
+    public void increaseDelivered() {
+        count.add(count.get(count.size() - 1) + 1);
+        //same deal with successCount as with failCount
+        double last = successCount.get(successCount.size() - 1);
+        successCount.add(last + 1);
+        failCount.add(failCount.get(failCount.size() - 1));
+        chart.updateXYSeries("Failed", count, failCount, null);
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: move this updating of series in private method and describe it, reuse it here

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        double[] initData = new double[1];
+        initData[0] = 0;
+        chart = QuickChart.getChart(title, xAxisName, yAxisName,"Failed", initData, initData);
+        chart.addSeries("Delivered", initData, initData);
+        chart.getSeriesMap().get("Failed").setLineColor(Color.RED);
+        chart.getSeriesMap().get("Failed").setMarker(new org.knowm.xchart.style.markers.None());
+        chart.getSeriesMap().get("Delivered").setLineColor(Color.green);
+        chart.getSeriesMap().get("Delivered").setMarker(new org.knowm.xchart.style.markers.None());
+        chartPanel = new XChartPanel(chart);
+        successCount = new CopyOnWriteArrayList<>();
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: move this to private method and name it  initCounters or something like that

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+        failCount = new CopyOnWriteArrayList<>();
+        failCount.add((double)0);
+        count = new CopyOnWriteArrayList<>();
+        count.add((double)0);
+        chart.updateXYSeries("Failed", count, failCount, null);
+        chart.updateXYSeries("Delivered", count, successCount, null);
+        chartPanel.repaint();
+    }
+
+}
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: I saw that MobileClient fat jar is added in src/main/resources. Remove this file.

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ *
+ * @author User
+ */
+public class GraphPanel {
+
+    private XYChart chart;
+    private XChartPanel chartPanel;
+    List<Double> successCount;
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/GraphPanel.java]: make this variables private

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaEidResultsGraphPanel.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+import org.knowm.xchart.XYChart;
+
+public class FrejaEidResultsGraphPanel {
+    private XYChart chart;
+    private XChartPanel chartPanel;
+    private List<Double> successCount;
+    private List<Double> failCount;
+    private List<Double> totalCount;
+    
+    private void resetLists(){
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaEidResultsGraphPanel.java]: Move private methods below constructor

File: src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaEidResultsGraphPanel.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        totalCount = new CopyOnWriteArrayList<>();
+        totalCount.add((double)0);
+    }
+    
+    private void updateGraph(){
+        chart.updateXYSeries("Failed", totalCount, failCount, null);
+        chart.updateXYSeries("Delivered", totalCount, successCount, null);
+    }
+
+    private void chartStyle(Color failColor, Color successColor){
```
Comment (aleksandarmarkovic-verisec) [inline src/main/java/com/internship/frejaeidjmeterplugin/jmeter/visualizer/gui/FrejaEidResultsGraphPanel.java]: rename to setChartStyle

**Review by aleksandarmarkovic-verisec — APPROVED** [review]

**Pull request merged**